src/main/java/org/apache/sling/models/impl/ModelAdapterFactory.java

```diff
@@ -404,11 +430,7 @@ public boolean isModelClass(@NotNull Class<?> type) {
                         ModelType model = (ModelType) Proxy.newProxyInstance(modelClass.getType().getClassLoader(), new Class<?>[] { modelClass.getType() }, handlerResult.getValue());
 
                         if (modelAnnotation.cache()) {
-                            Map<Class<?>, SoftReference<Object>> adaptableCache = adapterCache.get(cacheKey);
-                            if (adaptableCache == null) {
-                                adaptableCache = Collections.synchronizedMap(new WeakHashMap<Class<?>, SoftReference<Object>>());
-                                adapterCache.put(cacheKey, adaptableCache);
-                            }
+                            Map<Class<?>, SoftReference<Object>> adaptableCache = getOrCreateCache(adaptable);
```

Comment:
Don't we already know here that there was no cache created for this `adaptable`? If a cache was available, we would have returned the result already - see https://github.com/apache/sling-org-apache-sling-models-impl/pull/21/files#diff-0c273224a9fa6e263f4381c5d4b8145aa2c6c89e75e0b62025eea5efa17edec9R401.

Reply:
No there could be a cache with the same `adaptable` but a different "target" model class. For example, one could first `adapt` the request to `MyModel.class` and then to `MyOtherModel.class`. So in the first case, the cache doesn't exist and the model is of course not cached, but in the 2nd case, the cache exists but the model is also not cached.

Reply:
No, the same adaptable can be adapted to multiple types. For example, the request could be adapted first to `MyFirstModel.class` and then to `MySecondModel.class`: so in the first case there is no cache and of course no cached model, but in the 2nd case, the cache exists but doesn't contain a cached model for the `MySecondModel.class` (but would contain a cached model for `MyFirstModel.class`).

Reply:
I think my question was not clear enough. I'm saying that if the `modelAnnotation#cache` returns `true`, we can already create the cache for the `adaptable` at https://github.com/apache/sling-org-apache-sling-models-impl/pull/21/files#diff-0c273224a9fa6e263f4381c5d4b8145aa2c6c89e75e0b62025eea5efa17edec9R401, using `getOrCreate`.

Further down, you just need to populate the cache. I think the `getCache` and `getOrCreate` should be merged and make the code simpler to read.